#Theorical questions

##**Question 1: What is K-Nearest Neighbors (KNN) and how does it work in both
classification and regression problems?**


Ans- KNN works on the idea that similar data points exist close to each other. It does not learn a model explicitly; instead, it stores the training data and makes predictions based on similarity (distance).


**How KNN Works (Step-by-Step)**

- Choose the number of neighbors (K).
- Calculate the distance between the new data point and all training data points (commonly using Euclidean distance).
- Identify the K nearest neighbors.
- Make prediction based on these neighbors.

**KNN in Classification**

- Used when the output is a category (label).
- The algorithm assigns the class that is most common among the K nearest neighbors.

**Example**

If K = 3 and among nearest neighbors:

2 belong to Class A
1 belongs to Class B

➡️ Output = Class A (majority voting)

##**Question 2: What is the Curse of Dimensionality and how does it affect KNN
performance?**

The **Curse of Dimensionality** refers to the problems that arise when working with data that has a very large number of features (dimensions). As the number of dimensions increases, the amount of data needed to accurately analyze patterns grows exponentially, making many machine learning algorithms less effective.

##Understanding the Curse of Dimensionality

Imagine you have a dataset with only one feature, such as age. Finding similar data points is relatively easy. However, when the dataset contains dozens or hundreds of features, the data points become increasingly spread out in the feature space.

As dimensionality increases:

* The distance between data points becomes larger.
* Data becomes sparse (less densely packed).
* Finding truly "nearest" neighbors becomes difficult.
* Computational complexity increases significantly.

For example:

* In a 2-dimensional space, points may be clustered closely.
* In a 100-dimensional space, almost all points tend to be far apart from each other.

##How the Curse of Dimensionality Affects KNN

The **K-Nearest Neighbors (KNN)** algorithm relies heavily on distance calculations to identify the nearest neighbors. The curse of dimensionality negatively impacts KNN in several ways:

###1. Distance Measures Become Less Meaningful

KNN uses distance metrics such as Euclidean distance. In high-dimensional spaces, the difference between the nearest and farthest points becomes very small, making it difficult to distinguish which points are actually close.

Example:

**Low Dimensions:**

Nearest distance = 5

Farthest distance = 50


**High Dimensions:**

Nearest distance = 98

Farthest distance = 100

As a result, all points appear almost equally distant.

##2. Reduced Prediction Accuracy

Because KNN cannot effectively identify meaningful neighbors in high-dimensional data, its classification and prediction accuracy decreases.

##3. Increased Computational Cost

KNN calculates distances from the query point to every training sample. With more dimensions:

* More calculations are required.
* Memory consumption increases.
* Prediction time becomes slower.

##4. Requires Much More Training Data

As dimensionality grows, exponentially more data is needed to maintain the same level of accuracy. In many practical situations, collecting such large datasets is not feasible.

###Example

Suppose we are using KNN to classify customers:

| Features                 | Performance           |
| ------------------------ | --------------------- |
| Age, Salary (2 features) | High accuracy         |
| 100 customer attributes  | Lower accuracy        |
| 1000 attributes          | Very poor performance |

With 1000 features, most customers appear equally distant from each other, making neighbor selection unreliable.

##Methods to Reduce the Curse of Dimensionality

Several techniques can help improve KNN performance:

* **Feature Selection:** Remove irrelevant or redundant features.
* **Dimensionality Reduction:** Use techniques like **Principal Component Analysis (PCA).**
* **Feature Engineering:** Create more informative features.
* **Normalization/Standardization:** Scale features appropriately.
* **Use alternative algorithms:** Consider algorithms less sensitive to high dimensionality.

##Conclusion

The **Curse of Dimensionality** is a phenomenon where high-dimensional data becomes sparse and distance measurements lose their effectiveness. Since KNN depends entirely on distance calculations, increasing the number of features can significantly reduce its accuracy, increase computational cost, and require much larger datasets. Therefore, dimensionality reduction and feature selection are often essential before applying KNN to high-dimensional data.

## **Question 3: What is Principal Component Analysis (PCA)? How is it different from Feature Selection?**

**Principal Component Analysis (PCA)** is a dimensionality reduction technique used to reduce the number of features in a dataset while preserving as much information (variance) as possible. PCA transforms the original features into a new set of uncorrelated variables called **principal components.**

The first principal component captures the maximum variance in the data, the second captures the next highest variance, and so on. By selecting only the most important principal components, PCA reduces the dimensionality of the dataset while retaining most of the useful information.

How PCA Works
* PCA analyzes the relationships between features.
* It creates new features called principal components.
* These components are linear combinations of the original features.
* The components are ordered according to the amount of variance they explain.
* Only the top components are selected to reduce dimensionality.

###Difference Between PCA and Feature Selection
| Feature Selection                                                   | Principal Component Analysis (PCA)                                        |
| ------------------------------------------------------------------- | ------------------------------------------------------------------------- |
| Selects a subset of the original features.                          | Creates new features (principal components).                              |
| Original features remain unchanged.                                 | Original features are transformed into new variables.                     |
| Maintains interpretability of features.                             | Reduces interpretability because components are combinations of features. |
| Removes irrelevant or redundant features.                           | Combines features to preserve maximum variance.                           |
| Examples: Forward Selection, Backward Elimination, Chi-square test. | Example: PCA algorithm.                                                   |

###Example

Suppose a dataset contains the following features:

* Age
* Height
* Weight
* Income
* Education Level

###Feature Selection:

Feature selection might choose only:

* Height
* Weight
* Income

###PCA:

PCA would create new variables such as:

* PC1 = 0.5(Height) + 0.4(Weight) + 0.7(Income)
* PC2 = another combination of the original features

These new components capture most of the information present in the original dataset.

##Conclusion

**Principal Component Analysis (PCA)** is a dimensionality reduction technique that transforms the original features into a smaller set of principal components while preserving maximum variance. Unlike **feature selection**, which chooses existing features, PCA creates entirely new features by combining the original variables.

## **Question 4: What are eigenvalues and eigenvectors in PCA, and why are they important?**

In **Principal Component Analysis (PCA), eigenvectors** and eigenvalues are mathematical concepts used to identify the most important directions and the amount of information present in the data.

###Eigenvectors

**Eigenvectors** represent the directions (principal components) along which the data varies the most. They determine the new axes onto which the original data is projected.

In PCA:

* Each eigenvector corresponds to a principal component.
* Eigenvectors define the direction of maximum variance in the dataset.
* They are used to transform the original data into a lower-dimensional space.

###Eigenvalues

**Eigenvalues** represent the amount of variance or information captured by their corresponding eigenvectors.

In PCA:

* A larger eigenvalue indicates that the corresponding principal component explains more variance in the data.
* A smaller eigenvalue indicates less important information.

###Importance of Eigenvalues and Eigenvectors in PCA

Eigenvalues and eigenvectors are important because they help PCA determine:

* Which directions contain the most information (using eigenvectors).
* How much information each direction contains (using eigenvalues).
* Which principal components should be retained for dimensionality reduction.

Typically, PCA keeps the components with the largest eigenvalues and discards those with smaller eigenvalues.

###Example

Suppose PCA produces the following results:

| Principal Component | Eigenvalue | Variance Explained |
| ------------------- | ---------- | ------------------ |
| PC1                 | 8.5        | 70%                |
| PC2                 | 2.5        | 20%                |
| PC3                 | 0.8        | 7%                 |
| PC4                 | 0.2        | 3%                 |

In this case:

* PC1 and PC2 explain 90% of the total variance.
* Therefore, PCA may retain only these two components and discard the remaining ones.

###Simple Analogy

Imagine a cloud of data points spread across space:

* Eigenvectors determine the best directions to view the data.
* Eigenvalues tell us how much information is visible from each direction.

##Conclusion

In PCA, **eigenvectors** define the directions of the principal components, while **eigenvalues** measure the amount of variance captured by those components. They are essential because they enable PCA to identify and retain the most informative features while reducing the dimensionality of the dataset.

## **Question 5: How do KNN and PCA complement each other when applied in a single pipeline?**

**K-Nearest Neighbors (KNN)** and **Principal Component Analysis (PCA)** complement each other because **PCA reduces the dimensionality of the data,** while **KNN uses the reduced data to perform more efficient and accurate classification or prediction.**

KNN is highly sensitive to high-dimensional data due to the **curse of dimensionality.** PCA helps overcome this problem by reducing the number of features while preserving most of the important information. When used together in a single pipeline, PCA improves the performance of KNN.

###How PCA Helps KNN

#### **1. Reduces Dimensionality**

PCA transforms the original dataset into a smaller number of principal components, reducing the number of features that KNN must process.

#### **2. Improves Distance Calculations**

KNN relies on distance metrics such as Euclidean distance. PCA removes redundant and less important features, making distance measurements more meaningful.

#### **3. Reduces Computational Cost**

Since KNN calculates distances to all training samples, reducing the number of features decreases computation time and memory usage.

#### **4. Helps Overcome the Curse of Dimensionality**

By reducing the feature space, PCA makes data points less sparse and improves the effectiveness of nearest-neighbor searches.

#### **5. Can Improve Prediction Accuracy**

Removing noisy and irrelevant information often allows KNN to identify more reliable neighbors, leading to better classification performance.

###KNN-PCA Pipeline

The typical workflow is:

Original Dataset

↓

Data Preprocessing (Scaling/Normalization)

↓

Principal Component Analysis (PCA)

↓

Reduced-Dimension Dataset

↓

K-Nearest Neighbors (KNN)

↓

Prediction/Classification

###Example

Suppose a dataset has 100 features:

* Applying KNN directly on 100 features may result in:
* High computation time
* Poor distance measurements
* Lower accuracy
* After applying PCA and reducing the dataset to 10 principal components:
* Computation becomes faster
* Distance calculations become more meaningful
* KNN performance often improves

Advantages of Combining PCA and KNN

* Reduces the number of features
* Improves computational efficiency
* Reduces noise and redundancy
* Mitigates the curse of dimensionality
* Often increases classification accuracy
* Requires less memory

##Conclusion

**PCA and KNN complement each other because PCA reduces the dimensionality and complexity of the data, while KNN performs classification on the reduced feature space.** This combination helps improve KNN's speed, efficiency, and accuracy, especially when working with high-dimensional datasets.

## **Question 6: Train a KNN Classifier on the Wine Dataset with and without Feature Scaling. Compare Model Accuracy in Both Cases.**

## **Python Code**

In [ ]:
# Import libraries
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# Load Wine dataset
wine = load_wine()
X = wine.data
y = wine.target

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# -----------------------------
# KNN without Feature Scaling
# -----------------------------
knn_without_scaling = KNeighborsClassifier(n_neighbors=5)
knn_without_scaling.fit(X_train, y_train)

y_pred_without = knn_without_scaling.predict(X_test)

accuracy_without = accuracy_score(y_test, y_pred_without)

print("Accuracy without Feature Scaling:",
      round(accuracy_without * 100, 2), "%")


# -----------------------------
# KNN with Feature Scaling
# -----------------------------
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

knn_with_scaling = KNeighborsClassifier(n_neighbors=5)
knn_with_scaling.fit(X_train_scaled, y_train)

y_pred_with = knn_with_scaling.predict(X_test_scaled)

accuracy_with = accuracy_score(y_test, y_pred_with)

print("Accuracy with Feature Scaling:",
      round(accuracy_with * 100, 2), "%")

Accuracy without Feature Scaling: 72.22 %
Accuracy with Feature Scaling: 94.44 %


## **Output**

Accuracy without Feature Scaling: 72.22 %

Accuracy with Feature Scaling: 94.44 %

## **Comparison**

| Model                       | Accuracy |
| --------------------------- | -------- |
| KNN without Feature Scaling | 72.22%   |
| KNN with Feature Scaling    | 94.44%   |

##Conclusion

The KNN classifier performed significantly better after applying **feature scaling**. This is because KNN relies on distance calculations, and features with larger numerical values can dominate the distance metric if the data is not scaled. Using **StandardScaler** ensures that all features contribute equally, resulting in improved model performance and higher accuracy.

## **Question 7: Train a PCA model on the Wine dataset and print the explained variance ratio of each principal component.**

### **Python Code**

In [ ]:
# Import libraries
from sklearn.datasets import load_wine
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import pandas as pd

# Load Wine dataset
wine = load_wine()
X = wine.data

# Standardize the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Apply PCA
pca = PCA()
X_pca = pca.fit_transform(X_scaled)

# Print explained variance ratio
print("Explained Variance Ratio of Each Principal Component:\n")

for i, variance in enumerate(pca.explained_variance_ratio_):
    print(f"PC{i+1}: {variance:.4f}")

Explained Variance Ratio of Each Principal Component:

PC1: 0.3620
PC2: 0.1921
PC3: 0.1112
PC4: 0.0707
PC5: 0.0656
PC6: 0.0494
PC7: 0.0424
PC8: 0.0268
PC9: 0.0222
PC10: 0.0193
PC11: 0.0174
PC12: 0.0130
PC13: 0.0080


### **Output**

Explained Variance Ratio of Each Principal Component:

PC1: 0.3620

PC2: 0.1921

PC3: 0.1112

PC4: 0.0707

PC5: 0.0656

PC6: 0.0494

PC7: 0.0424

PC8: 0.0268

PC9: 0.0222

PC10: 0.0193

PC11: 0.0174

PC12: 0.0129

PC13: 0.0079

###Interpretation

* **PC1** explains approximately **36.20%** of the total variance.
* **PC2** explains approximately **19.21%** of the total variance.
* Together, **PC1 and PC2** explain about **55.41%** of the total variance.
* The first few principal components capture most of the information in the dataset, allowing us to reduce dimensionality while preserving the majority of the variance.

##Conclusion

The PCA model transforms the original 13 features of the Wine dataset into 13 principal components ranked by the amount of variance they explain. The explained variance ratio helps determine how many principal components should be retained for dimensionality reduction while minimizing information loss.

## **Question 8: Train a KNN Classifier on the PCA-transformed dataset (retain top 2 components). Compare the accuracy with the original dataset.**

### **Python Code**

In [ ]:
# Import libraries
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Load Wine dataset
wine = load_wine()
X = wine.data
y = wine.target

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# ----------------------------------
# KNN on Original Dataset
# ----------------------------------
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

knn_original = KNeighborsClassifier(n_neighbors=5)
knn_original.fit(X_train_scaled, y_train)

y_pred_original = knn_original.predict(X_test_scaled)

original_accuracy = accuracy_score(y_test, y_pred_original)

# ----------------------------------
# PCA Transformation (2 Components)
# ----------------------------------
pca = PCA(n_components=2)

X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

# KNN on PCA-transformed Dataset
knn_pca = KNeighborsClassifier(n_neighbors=5)
knn_pca.fit(X_train_pca, y_train)

y_pred_pca = knn_pca.predict(X_test_pca)

pca_accuracy = accuracy_score(y_test, y_pred_pca)

# Print results
print("Accuracy on Original Dataset:",
      round(original_accuracy * 100, 2), "%")

print("Accuracy on PCA Dataset (2 Components):",
      round(pca_accuracy * 100, 2), "%")

Accuracy on Original Dataset: 94.44 %
Accuracy on PCA Dataset (2 Components): 100.0 %


### **Output**

Accuracy on Original Dataset: 94.44 %

Accuracy on PCA Dataset (2 Components): 100.0 %

### **Comparison**

| Dataset                 | Number of Features | Accuracy |
| ----------------------- | ------------------ | -------- |
| Original Scaled Dataset | 13                 | 94.44%   |
| PCA Dataset             | 2                  | 100.0%   |

## **Conclusion**

In this experiment, the KNN classifier achieved an accuracy of **94.44%** on the original scaled Wine dataset. After applying **PCA** and retaining only the **top 2 principal components**, the model achieved an accuracy of **100.00%.** This indicates that PCA successfully reduced the dimensionality of the dataset while preserving the most important information, resulting in improved KNN performance and reduced computational complexity.

## **Question 9: Train a KNN Classifier with different distance metrics (Euclidean, Manhattan) on the scaled Wine dataset and compare the results.**

## **Python Code**

In [ ]:
# Import libraries
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Load Wine dataset
wine = load_wine()
X = wine.data
y = wine.target

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Feature Scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# -----------------------------------
# KNN with Euclidean Distance
# -----------------------------------
knn_euclidean = KNeighborsClassifier(
    n_neighbors=5,
    metric='euclidean'
)

knn_euclidean.fit(X_train_scaled, y_train)
y_pred_euclidean = knn_euclidean.predict(X_test_scaled)

euclidean_accuracy = accuracy_score(
    y_test,
    y_pred_euclidean
)

# -----------------------------------
# KNN with Manhattan Distance
# -----------------------------------
knn_manhattan = KNeighborsClassifier(
    n_neighbors=5,
    metric='manhattan'
)

knn_manhattan.fit(X_train_scaled, y_train)
y_pred_manhattan = knn_manhattan.predict(X_test_scaled)

manhattan_accuracy = accuracy_score(
    y_test,
    y_pred_manhattan
)

# Print Results
print("Euclidean Distance Accuracy:",
      round(euclidean_accuracy * 100, 2), "%")

print("Manhattan Distance Accuracy:",
      round(manhattan_accuracy * 100, 2), "%")

Euclidean Distance Accuracy: 94.44 %
Manhattan Distance Accuracy: 94.44 %


## **Output**

Euclidean Distance Accuracy: 94.44 %

Manhattan Distance Accuracy: 94.44 %

## **Comparison**

| Distance Metric | Accuracy |
| --------------- | -------- |
| Euclidean       | 94.44%   |
| Manhattan       | 94.44%   |

## **Conclusion**

In this experiment, the KNN classifier was trained using two different distance metrics: **Euclidean** and **Manhattan**. After applying feature scaling to the Wine dataset, both distance metrics achieved the same classification accuracy of **94.44%.** This indicates that, for this particular dataset and value of k, the choice between Euclidean and Manhattan distance did not significantly affect the performance of the KNN classifier. However, the effectiveness of a distance metric may vary depending on the characteristics of the dataset.

## **Question 10: You are working with a high-dimensional gene expression dataset to classify patients with different types of cancer. Explain how you would use PCA and KNN to build a robust classification pipeline.**

In a high-dimensional gene expression dataset, the number of features (genes) is often much larger than the number of patient samples. This can lead to overfitting, increased computational cost, and poor model generalization. To address these challenges, I would use a **PCA + KNN pipeline.**

### **1. Use PCA to Reduce Dimensionality**

I would first standardize the gene expression data using **StandardScaler** because PCA is sensitive to feature scales. Then, I would apply **Principal Component Analysis (PCA)** to transform the original high-dimensional gene features into a smaller set of principal components that retain most of the variance in the data.

####Benefits of PCA include:

* Reducing the number of features
* Removing redundant and noisy information
* Mitigating the curse of dimensionality
* Reducing computational complexity
* Improving model generalization

###2. Decide How Many Components to Keep

To determine the optimal number of principal components, I would examine the **cumulative explained variance ratio** and select the minimum number of components that preserve approximately **90%–95% of the total variance.**

For example:

Component 1: 40%

Component 2: 25%

Component 3: 15%

Component 4: 10%

Total Explained Variance: 90%

In this case, the first four principal components would be retained.

###3. Use KNN for Classification After PCA

After dimensionality reduction, I would train a **K-Nearest Neighbors (KNN)** classifier on the transformed dataset.

Steps include:

* Split the dataset into training and testing sets.
* Apply feature scaling.
* Perform PCA transformation.
* Train KNN using an optimal value of K determined through cross-validation.
* Predict cancer classes using the nearest neighbors in the reduced feature space.

###4. Evaluate the Model

The model would be evaluated using several performance metrics:

* Accuracy
* Precision
* Recall
* F1-score
* Confusion Matrix
* Cross-validation accuracy
* ROC-AUC score (when applicable)

Cross-validation is particularly important because biomedical datasets often have limited samples.

###5. Justification to Stakeholders

I would justify this pipeline as a robust solution because:

* **PCA reduces overfitting** by removing noise and redundant features.
* **KNN performs better** in lower-dimensional spaces.
* The pipeline reduces computational cost and memory requirements.
* Cross-validation ensures reliable performance estimation.
* The approach is widely used and validated in biomedical and genomics research.
* The model becomes more generalizable and suitable for real-world clinical applications.

This pipeline balances **accuracy, efficiency,** and **robustness,** making it appropriate for high-dimensional biomedical datasets.

### **Python Code**

In [ ]:
# Import libraries
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

# Load dataset (using Wine dataset as a sample)
wine = load_wine()
X = wine.data
y = wine.target

# Split data
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Feature scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Apply PCA (retain 95% variance)
pca = PCA(n_components=0.95)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

# Train KNN
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_pca, y_train)

# Prediction
y_pred = knn.predict(X_test_pca)

# Evaluation
print("Number of PCA Components:",
      pca.n_components_)

print("Accuracy:",
      round(accuracy_score(y_test, y_pred) * 100, 2), "%")

print("\nClassification Report:\n")
print(classification_report(y_test, y_pred))

Number of PCA Components: 10
Accuracy: 94.44 %

Classification Report:

              precision    recall  f1-score   support

           0       0.93      1.00      0.97        14
           1       1.00      0.86      0.92        14
           2       0.89      1.00      0.94         8

    accuracy                           0.94        36
   macro avg       0.94      0.95      0.94        36
weighted avg       0.95      0.94      0.94        36



## **Output**

Number of PCA Components: 10

Accuracy: 94.44 %

Classification Report:

              precision    recall  f1-score   support

           0       0.93      1.00      0.97        14
           1       1.00      0.86      0.92        14
           2       0.89      1.00      0.94         8

    accuracy                           0.94        36
   macro avg       0.94      0.95      0.94        36
weighted avg       0.95      0.94      0.94        36

###Interpretation

* PCA reduced the original feature space to **10 principal components** while retaining approximately **95% of the variance.**
* The KNN classifier achieved an overall **accuracy of 94.44%.**
* The model showed strong performance across all cancer classes, with **precision**, **recall**, and **F1-scores** above **0.90** for most classes.
* The use of PCA helped reduce dimensionality and mitigate overfitting while maintaining high classification accuracy.

##Conclusion

The **PCA + KNN pipeline** provides an effective solution for high-dimensional biomedical datasets such as gene expression data. PCA reduces the number of features by retaining only the most informative components, which helps overcome the curse of dimensionality and reduces the risk of overfitting. After dimensionality reduction, KNN performs classification efficiently in the reduced feature space.

In this experiment, PCA reduced the dataset to **10 principal components**, and the KNN classifier achieved an accuracy of **94.44%.** The high precision, recall, and F1-scores demonstrate that this pipeline can provide robust and reliable classification performance. Therefore, the **PCA + KNN** approach is a practical and scalable solution for real-world biomedical applications, where datasets often contain a large number of features but relatively few samples.